File: src/lib/health-data.ts
import type { MetricDay } from './metric-series';
import { getMetricDef } from './metric-catalog';

export type HealthMetric = {
  id: string;
  label: string;
  value: number | null;
  unit: string;
  status: 'loaded' | 'empty' | 'error';
  error?: string;
  dailyValues?: MetricDay[];
};

export type ExerciseSummary = {
  id: string;
  name: string;
  type: string;
  startTime?: string;
  endTime?: string;
  activeMinutes: number | null;
  caloriesKcal: number | null;
  distanceKm: number | null;
  steps: number | null;
};

export type SleepSummary = {
  id: string;
  /** 'nap' when the API flags the session as a nap (or it looks like one) */
  kind: 'sleep' | 'nap';
  startTime?: string;
  endTime?: string;
  minutesAsleep: number | null;
  minutesInSleepPeriod: number | null;
};

export type HealthSnapshot = {
  metrics: HealthMetric[];
  exercises: ExerciseSummary[];
  /** Sleep sessions in the range, most recent first */
  sleepSessions: SleepSummary[];
  rangeLabel: string;
  raw: {
    rollups: Record<string, unknown>;
    exercises: unknown;
    sleep: unknown;
  };
};

export type HealthSnapshotOptions = {
  days?: number;
  /** Metric ids to load; defaults to the default rings + cards */
  metricIds?: string[];
};


export function formatMetricValue(metric: HealthMetric) {
  if (metric.value === null) {
    return '--';
  }

  const digits = getMetricDef(metric.id)?.fractionDigits ?? 0;
  return metric.value.toLocaleString(undefined, {
    maximumFractionDigits: digits,
  });
}

File: ios-swift/OpenFit/Model/HealthData.swift
import Foundation

struct MetricDay: Hashable {
    var date: String          // "yyyy-MM-dd" local
    var value: Double?
}

enum HealthMetricStatus: String {
    case loaded, empty, error
}

struct HealthMetric: Identifiable {
    var id: String
    var label: String
    var value: Double?
    var unit: String
    var status: HealthMetricStatus
    var error: String?
    var dailyValues: [MetricDay]?
}

struct ExerciseSummary: Identifiable, Hashable {
    var id: String
    var name: String
    var type: String
    var startTime: String?
    var endTime: String?
    var activeMinutes: Double?
    var caloriesKcal: Double?
    var distanceKm: Double?
    var steps: Double?
}

struct SleepSummary: Identifiable, Hashable {
    var id: String
    var kind: String          // "sleep" | "nap"
    var startTime: String?
    var endTime: String?
    var minutesAsleep: Double?
    var minutesInSleepPeriod: Double?
}

struct HealthSnapshot {
    var metrics: [HealthMetric]
    var exercises: [ExerciseSummary]
    /// Sleep sessions in the range, most recent first.
    var sleepSessions: [SleepSummary]
    var rangeLabel: String
}

func formatMetricValue(_ metric: HealthMetric) -> String {
    guard let value = metric.value else { return "--" }
    let digits = MetricCatalog.def(for: metric.id)?.fractionDigits ?? 0
    return formatAmount(value, fractionDigits: digits)
}
